### PR TITLE
Use golangci-lint to find unused/dead code (Experimental!)

### PR DESCRIPTION
**Pull request by dims**
`verify-golangci-lint.sh` is an Experimental script. This is currently intended to be run by hand only. This puts us on a path to try things out using the newer `golangci-lint` and replace a bunch of linters we currently use to the consolidated tool.

This script should NOT run in any CI jobs needless to say!

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: hack/verify-golangci-lint.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+echo 'installing golangci-lint '
+pushd "${KUBE_ROOT}/hack/tools" >/dev/null
+  go install github.com/golangci/golangci-lint/cmd/golangci-lint
+popd >/dev/null
+
+cd "${KUBE_ROOT}"
+
+echo 'running golangci-lint '
+golangci-lint run \
+  --timeout 30m \
```
Comment (dims) [inline hack/verify-golangci-lint.sh:48]: see list of linters here:
https://golangci-lint.run/usage/linters/

**Comment by dims** [comment]
looks like a handful of things we can cleanup from the logs ( https://prow.k8s.io/log?container=test&id=1361709652714196992&job=pull-kubernetes-verify )

```
running golangci-lint 
pkg/volume/util/volumepathhandler/volume_path_handler.go:34:2: `statPath` is unused (deadcode)
	statPath              = "stat"
	^
pkg/util/iptables/iptables.go:632:2: `opAppendRule` is unused (deadcode)
	opAppendRule  operation = "-A"
	^
cluster/images/etcd/migrate/versions.go:129:6: `MustParseEtcdVersionPair` is unused (deadcode)
func MustParseEtcdVersionPair(s string) *EtcdVersionPair {
     ^
cluster/images/etcd/migrate/versions.go:191:6: `MustParseSupportedVersions` is unused (deadcode)
func MustParseSupportedVersions(list []string) SupportedVersions {
     ^
plugin/pkg/admission/imagepolicy/certs_test.go:24:5: `caKey` is unused (deadcode)
var caKey = []byte(`-----BEGIN RSA PRIVATE KEY-----
    ^
plugin/pkg/admission/imagepolicy/certs_test.go:72:5: `badCAKey` is unused (deadcode)
var badCAKey = []byte(`-----BEGIN RSA PRIVATE KEY-----
    ^
pkg/kubelet/certificate/bootstrap/bootstrap_test.go:144:2: `noError` is unused (deadcode)
	noError failureType = iota
	^
cmd/kube-proxy/app/server.go:94:2: `proxyModeKernelspace` is unused (deadcode)
	proxyModeKernelspace = "kernelspace"
	^
pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio.go:34:2: `minUtilization` is unused (deadcode)
	minUtilization               = 0
	^
pkg/volume/vsphere_volume/vsphere_volume_util.go:39:2: `maxRetries` is unused (deadcode)
	maxRetries         = 10
	^
pkg/volume/glusterfs/glusterfs.go:80:2: `durabilityType` is unused (deadcode)
	durabilityType                 = "replicate"
	^
pkg/controller/volume/attachdetach/reconciler/reconciler_test.go:42:2: `resyncPeriod` is unused (deadcode)
	resyncPeriod              time.Duration = 5 * time.Minute
	^
pkg/kubelet/dockershim/helpers_linux.go:159:6: `applyExperimentalCreateConfig` is unused (deadcode)
func applyExperimentalCreateConfig(createConfig *dockertypes.ContainerCreateConfig, annotations map[string]string) {
     ^
test/e2e/upgrades/apps/replicasets.go:37:2: `interval` is unused (deadcode)
	interval = 10 * time.Second
	^
test/e2e/upgrades/apps/replicasets.go:38:2: `timeout` is unused (deadcode)
	timeout  = 5 * time.Minute
	^
test/e2e/kubectl/kubectl.go:104:2: `httpdRCFilename` is unused (deadcode)
	httpdRCFilename           = "httpd-rc.yaml.in"
	^
test/e2e/kubectl/kubectl.go:118:2: `cronJobGroupVersionResourceAlpha` is unused (deadcode)
	cronJobGroupVersionResourceAlpha = schema.GroupVersionResource{Group: "batch", Version: "v2alpha1", Resource: "cronjobs"}
	^
test/e2e/kubectl/kubectl.go:119:2: `cronJobGroupVersionResourceBeta` is unused (deadcode)
	cronJobGroupVersionResourceBeta  = schema.GroupVersionResource{Group: "batch", Version: "v1beta1", Resource: "cronjobs"}
	^
test/e2e/storage/csi_mock_volume.go:1635:2: `podStartTimeout` is unused (deadcode)
	podStartTimeout                = 5 * time.Minute
	^
test/e2e/network/kube_proxy.go:45:3: `testDaemonHTTPPort` is unused (varcheck)
		testDaemonHTTPPort    = 11301
		^
test/e2e/apimachinery/flowcontrol.go:114:4: field `flowSchemaName` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:209:4: field `concurrencyMultiplier` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:99:2: field `barMap` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:98:2: field `bars` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:212:4: field `expectedCompletedPercentage` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:95:2: field `foo` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:90:2: field `strptr` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:111:4: field `priorityLevelName` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:92:2: field `ints` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:85:6: type `foo` is unused (unused)
pkg/controller/replicaset/replica_set_test.go:508:6: type `byName` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:112:4: field `concurrencyMultiplier` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:117:4: field `expectedCompletedPercentage` is unused (unused)
test/images/agnhost/net/nat/closewait.go:42:5: var `leakedConnection` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:115:4: field `matchingPrecedence` is unused (unused)
pkg/util/coverage/fake_test_deps.go:27:6: type `fakeTestDeps` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:89:2: field `str` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:96:2: field `fooptr` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:93:2: field `stringMap` is unused (unused)
```

**Comment by palnabarun** [comment]
@dims -- I am thinking to take up the `test/e2e/apimachinery/**` failures as a demo PR for the workshop this weekend. Would that be fine by you?

**Comment by dims** [comment]
@palnabarun sure

**Comment by siddhartha97** [comment]
@dims can i work on this? :)

File: hack/verify-golangci-lint.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
+
+# Copyright 2014 The Kubernetes Authors.
```
Comment (neolit123) [inline hack/verify-golangci-lint.sh:3]: 2021

**Comment by neolit123** [comment]
with the current scope the detected errors seem fine to fix:

<details>

```
running golangci-lint 
pkg/volume/util/volumepathhandler/volume_path_handler.go:34:2: `statPath` is unused (deadcode)
	statPath              = "stat"
	^
pkg/util/iptables/iptables.go:632:2: `opAppendRule` is unused (deadcode)
	opAppendRule  operation = "-A"
	^
cluster/images/etcd/migrate/versions.go:129:6: `MustParseEtcdVersionPair` is unused (deadcode)
func MustParseEtcdVersionPair(s string) *EtcdVersionPair {
     ^
cluster/images/etcd/migrate/versions.go:191:6: `MustParseSupportedVersions` is unused (deadcode)
func MustParseSupportedVersions(list []string) SupportedVersions {
     ^
plugin/pkg/admission/imagepolicy/certs_test.go:24:5: `caKey` is unused (deadcode)
var caKey = []byte(`-----BEGIN RSA PRIVATE KEY-----
    ^
plugin/pkg/admission/imagepolicy/certs_test.go:72:5: `badCAKey` is unused (deadcode)
var badCAKey = []byte(`-----BEGIN RSA PRIVATE KEY-----
    ^
pkg/kubelet/certificate/bootstrap/bootstrap_test.go:144:2: `noError` is unused (deadcode)
	noError failureType = iota
	^
cmd/kube-proxy/app/server.go:94:2: `proxyModeKernelspace` is unused (deadcode)
	proxyModeKernelspace = "kernelspace"
	^
pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio.go:34:2: `minUtilization` is unused (deadcode)
	minUtilization               = 0
	^
pkg/volume/vsphere_volume/vsphere_volume_util.go:39:2: `maxRetries` is unused (deadcode)
	maxRetries         = 10
	^
pkg/volume/glusterfs/glusterfs.go:80:2: `durabilityType` is unused (deadcode)
	durabilityType                 = "replicate"
	^
pkg/controller/volume/attachdetach/reconciler/reconciler_test.go:42:2: `resyncPeriod` is unused (deadcode)
	resyncPeriod              time.Duration = 5 * time.Minute
	^
pkg/kubelet/dockershim/helpers_linux.go:159:6: `applyExperimentalCreateConfig` is unused (deadcode)
func applyExperimentalCreateConfig(createConfig *dockertypes.ContainerCreateConfig, annotations map[string]string) {
     ^
test/e2e/upgrades/apps/replicasets.go:37:2: `interval` is unused (deadcode)
	interval = 10 * time.Second
	^
test/e2e/upgrades/apps/replicasets.go:38:2: `timeout` is unused (deadcode)
	timeout  = 5 * time.Minute
	^
test/e2e/kubectl/kubectl.go:104:2: `httpdRCFilename` is unused (deadcode)
	httpdRCFilename           = "httpd-rc.yaml.in"
	^
test/e2e/kubectl/kubectl.go:118:2: `cronJobGroupVersionResourceAlpha` is unused (deadcode)
	cronJobGroupVersionResourceAlpha = schema.GroupVersionResource{Group: "batch", Version: "v2alpha1", Resource: "cronjobs"}
	^
test/e2e/kubectl/kubectl.go:119:2: `cronJobGroupVersionResourceBeta` is unused (deadcode)
	cronJobGroupVersionResourceBeta  = schema.GroupVersionResource{Group: "batch", Version: "v1beta1", Resource: "cronjobs"}
	^
test/e2e/storage/csi_mock_volume.go:1635:2: `podStartTimeout` is unused (deadcode)
	podStartTimeout                = 5 * time.Minute
	^
test/e2e/network/kube_proxy.go:45:3: `testDaemonHTTPPort` is unused (varcheck)
		testDaemonHTTPPort    = 11301
		^
test/e2e/apimachinery/flowcontrol.go:114:4: field `flowSchemaName` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:209:4: field `concurrencyMultiplier` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:99:2: field `barMap` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:98:2: field `bars` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:212:4: field `expectedCompletedPercentage` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:95:2: field `foo` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:90:2: field `strptr` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:111:4: field `priorityLevelName` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:92:2: field `ints` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:85:6: type `foo` is unused (unused)
pkg/controller/replicaset/replica_set_test.go:508:6: type `byName` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:112:4: field `concurrencyMultiplier` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:117:4: field `expectedCompletedPercentage` is unused (unused)
test/images/agnhost/net/nat/closewait.go:42:5: var `leakedConnection` is unused (unused)
test/e2e/apimachinery/flowcontrol.go:115:4: field `matchingPrecedence` is unused (unused)
pkg/util/coverage/fake_test_deps.go:27:6: type `fakeTestDeps` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:89:2: field `str` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:96:2: field `fooptr` is unused (unused)
pkg/kubelet/apis/config/helpers_test.go:93:2: field `stringMap` is unused (unused)
+++ exit code: 1
```

</details>

File: hack/verify-golangci-lint.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script checks coding style for go language files in each
+# Kubernetes package by golint.
+# Usage: `hack/verify-golint.sh`.
```
Comment (liggitt) [inline hack/verify-golangci-lint.sh:19]: nit: fix script name

File: hack/verify-golangci-lint.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+echo 'installing golangci-lint '
+pushd "${KUBE_ROOT}/hack/tools" >/dev/null
+  go install github.com/golangci/golangci-lint/cmd/golangci-lint
+popd >/dev/null
+
+cd "${KUBE_ROOT}"
+
+echo 'running golangci-lint '
+golangci-lint run \
+  --timeout 30m \
```
Comment (liggitt) [inline hack/verify-golangci-lint.sh:48]: what happens on timeout? are we close to this limit?

**Review by liggitt — COMMENTED** [review]
a couple questions, and looks like the verify failures are legitimate, but lgtm otherwise

File: hack/verify-golangci-lint.sh (hunk 3 above)
Comment (dims) [reply hack/verify-golangci-lint.sh:19]: on it!

File: hack/verify-golangci-lint.sh (hunk 4 above)
Comment (dims) [reply hack/verify-golangci-lint.sh:48]: the first run took
```[0;31mFAILED[0m   verify-golangci-lint.sh	365s```

it fails hard with an error ( example https://github.com/ovn-org/ovn-kubernetes/issues/1222 )

**Review by liggitt — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**